File: python-algorithm/ds_5_tree/tree_1_2_inter_recurrent/L687.py
#--coding:utf-8--


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    res = 0
    def longestUnivaluePath(self, root: TreeNode) -> int:
        if root == None:
            return 0
        self.cal_path(root)
        return self.res

    def cal_path(self, root):
        if root == None:
            return 0
        l_path = self.cal_path(root.left)
        r_path = self.cal_path(root.right)
        if root.left and root.left.val == root.val:
            l_num = l_path+1
        else:
            l_num = 0
        if root.right and root.right.val == root.val:
            r_num = r_path+1
        else:
            r_num = 0
        self.res = max(self.res, l_num+r_num)
        return max(l_num, r_num)

File: python-algorithm/ds_5_tree/tree_1_2_inter_recurrent/L687.cpp


//687. 最长同值路径

#include <iostream>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int res = 0;
    int get(TreeNode* root) {
        if (root == NULL) return 0;
        //这里l,r代表的是每个当前节点的左右子树的连续层个数,和113 543 两道题有区别
        int left = get(root->left);
        int right = get(root->right);
        //每个节点，都会做一次比较
        int leftPath = root->left != NULL && root->left->val == root->val? left+1 : 0;
        int rightPath = root->right != NULL && root->right->val == root->val? right+1 : 0;
        res = max(res, leftPath + rightPath);
        return max(leftPath, rightPath);
    }

    int longestUnivaluePath(TreeNode* root) {
        get(root);
        return res;
    }

};